Move more methods from math/engine to global namespace. Also s/scope/tidy/

dsmilkov opened the pull request:
In addition fix test in backend_cpu regarding `reverse`, so Travis passes.

<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/pair-code/deeplearnjs/625)
<!-- Reviewable:end -->

nsthorat commented on src/math/backends/gradients.ts:
```diff
@@ -0,0 +1,216 @@
+    },
+    inputs: NamedArrayMap): T {
+  name = name || '';
+  ENV.engine.startCustomGradient();
```
let's push all of this into the engine

nsthorat commented on src/math/backends/gradients.ts:
```diff
@@ -0,0 +1,216 @@
+    varList = varList.filter(variable => variable.trainable);
+  }
+
+  const gradientsMode = true;
```
let's push here and below into the engine. this in-function input computation should be hidden in the sugar floating function

nsthorat commented on src/math/backends/gradients.ts:
```diff
@@ -0,0 +1,216 @@
+  const keys = x instanceof NDArray ? null : Object.keys(x);
+  const xs = util.flattenNameArrayMap(x, keys);
+
+  const gradientsMode = true;
```
let's push here and below into engine

nsthorat commented on src/math/backends/gradients.ts:
```diff
@@ -0,0 +1,216 @@
+ *     using the provided name.
+ * @param scopeFn The function to execute.
+ */
+export function gradientsScope<T extends TidyResult>(
```
here we have "Scope" - see previous comment about "tidy" vs "scope" terminology

dsmilkov commented:
Review status: 0 of 71 files reviewed at latest revision, 26 unresolved discussions, some commit checks failed.

---

*[demos/latent-space-explorer/utils/FontModel.ts, line 62 at r1](https://beta.reviewable.io/reviews/pair-code/deeplearnjs/625#-L4DyQmN-cpAKxEFZK4s-r1-62:-L4J-D7PssXmDSw3tETJ:b-896fix) ([raw file](https://github.com/pair-code/deeplearnjs/blob/42d15710993c81e9af9bca5e81011203d4b245a9/demos/latent-space-explorer/utils/FontModel.ts#L62)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

Remove this
</blockquote></details>

Done.

---

*[docs/tutorials/ml_beginners.md, line 252 at r1](https://beta.reviewable.io/reviews/pair-code/deeplearnjs/625#-L4DyQmN-cpAKxEFZK51-r1-252:-L4J01mhE5Wkxrtxnxq8:b-896fix) ([raw file](https://github.com/pair-code/deeplearnjs/blob/42d15710993c81e9af9bca5e81011203d4b245a9/docs/tutorials/ml_beginners.md#L252)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

s/scope/tidy
</blockquote></details>

Done.

---

*[docs/tutorials/performance.md, line 74 at r1](https://beta.reviewable.io/reviews/pair-code/deeplearnjs/625#-L4DyQmN-cpAKxEFZK52-r1-74:-L4J2KWl8v5HFrJI3_8c:b-lq2ktq) ([raw file](https://github.com/pair-code/deeplearnjs/blob/42d15710993c81e9af9bca5e81011203d4b245a9/docs/tutorials/performance.md#L74)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

s/math.scope/dl.tidy
</blockquote></details>

Done. Also updated the tutorial

---

*[docs/tutorials/performance.md, line 92 at r1](https://beta.reviewable.io/reviews/pair-code/deeplearnjs/625#-L4DyQmN-cpAKxEFZK52-r1-92:-L4J2OA3hBYHI3xe9xrp:b-896fix) ([raw file](https://github.com/pair-code/deeplearnjs/blob/42d15710993c81e9af9bca5e81011203d4b245a9/docs/tutorials/performance.md#L92)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

dl.tidy
</blockquote></details>

Done.

---

*[docs/tutorials/performance.md, line 100 at r1](https://beta.reviewable.io/reviews/pair-code/deeplearnjs/625#-L4DyQmN-cpAKxEFZK52-r1-100:-L4J2OTpCyd9KNZDa17t:b-896fix) ([raw file](https://github.com/pair-code/deeplearnjs/blob/42d15710993c81e9af9bca5e81011203d4b245a9/docs/tutorials/performance.md#L100)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

dl.tidy
</blockquote></details>

Done.

---

*[docs/tutorials/performance.md, line 127 at r1](https://beta.reviewable.io/reviews/pair-code/deeplearnjs/625#-L4DyQmN-cpAKxEFZK52-r1-127:-L4J2OpPg_pWSXKTOY2u:b-896fix) ([raw file](https://github.com/pair-code/deeplearnjs/blob/42d15710993c81e9af9bca5e81011203d4b245a9/docs/tutorials/performance.md#L127)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

dl.tidy
</blockquote></details>

Done.

---

*[docs/tutorials/performance.md, line 132 at r1](https://beta.reviewable.io/reviews/pair-code/deeplearnjs/625#-L4DyQmN-cpAKxEFZK52-r1-132:-L4J2P5o5cE3iRSSUwip:b-896fix) ([raw file](https://github.com/pair-code/deeplearnjs/blob/42d15710993c81e9af9bca5e81011203d4b245a9/docs/tutorials/performance.md#L132)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

dl.tidy
</blockquote></details>

Done.

---

*[src/environment.ts, line 346 at r1](https://beta.reviewable.io/reviews/pair-code/deeplearnjs/625#-L4CoAMG9TgzKj0hu5Ss-r1-346:-L4J3MrVE1DlEQpVTLeG:b-896fix) ([raw file](https://github.com/pair-code/deeplearnjs/blob/42d15710993c81e9af9bca5e81011203d4b245a9/src/environment.ts#L346)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

move this up with other properties
</blockquote></details>

Done.

---

*[src/graph/session.ts, line 105 at r1](https://beta.reviewable.io/reviews/pair-code/deeplearnjs/625#-L4DyQmN-cpAKxEFZK5c-r1-105:-L4J3vw9zEH1x5XP1vPR:bsxt6bn) ([raw file](https://github.com/pair-code/deeplearnjs/blob/42d15710993c81e9af9bca5e81011203d4b245a9/src/graph/session.ts#L105)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

dl.tidy - I would mass find "math.tidy" so you also catch comments
</blockquote></details>

Done. Mass replaced

---

*[src/math/math.ts, line 200 at r1](https://beta.reviewable.io/reviews/pair-code/deeplearnjs/625#-L4CoAMG9TgzKj0hu5Su-r1-200:-L4J445kHQAZT5aYD-lp:b-896fix) ([raw file](https://github.com/pair-code/deeplearnjs/blob/42d15710993c81e9af9bca5e81011203d4b245a9/src/math/math.ts#L200)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

let's remove this. once you do, fix the fallout (where you call math.tidy in some places)
</blockquote></details>

Done.

---

*[src/math/backends/backend_engine.ts, line 33 at r1](https://beta.reviewable.io/reviews/pair-code/deeplearnjs/625#-L4CoAMG9TgzKj0hu5St-r1-33:-L4J46_3-thbYrXTVGy-:bcyu0iz) ([raw file](https://github.com/pair-code/deeplearnjs/blob/42d15710993c81e9af9bca5e81011203d4b245a9/src/math/backends/backend_engine.ts#L33)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

TidyState?
</blockquote></details>

Talked offline. Decided to keep it as scope internally. tidy is just a sugar facing the user.

---

*[src/math/backends/backend_engine.ts, line 58 at r1](https://beta.reviewable.io/reviews/pair-code/deeplearnjs/625#-L4CoAMG9TgzKj0hu5St-r1-58:-L4J4SGSAbGNCOddW8U2:b-896fix) ([raw file](https://github.com/pair-code/deeplearnjs/blob/42d15710993c81e9af9bca5e81011203d4b245a9/src/math/backends/backend_engine.ts#L58)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

maybe move this to the top since it's public
</blockquote></details>

Done.

---

*[src/math/backends/backend_engine.ts, line 107 at r1](https://beta.reviewable.io/reviews/pair-code/deeplearnjs/625#-L4CoAMG9TgzKj0hu5St-r1-107:-L4J4Y4K23pLVJnFE5_m:b-2exn2y) ([raw file](https://github.com/pair-code/deeplearnjs/blob/42d15710993c81e9af9bca5e81011203d4b245a9/src/math/backends/backend_engine.ts#L107)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

update this jsdoc
</blockquote></details>

This is internal private method. Kept it since we decided to keep it as scope internally.

---

*[src/math/backends/backend_engine.ts, line 118 at r1](https://beta.reviewable.io/reviews/pair-code/deeplearnjs/625#-L4CoAMG9TgzKj0hu5St-r1-118:-L4J5-EIQHUu-acGiOaZ:b-896fix) ([raw file](https://github.com/pair-code/deeplearnjs/blob/42d15710993c81e9af9bca5e81011203d4b245a9/src/math/backends/backend_engine.ts#L118)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

reminder to update this error
</blockquote></details>

Done.

---

*[src/math/backends/backend_engine.ts, line 174 at r1](https://beta.reviewable.io/reviews/pair-code/deeplearnjs/625#-L4CoAMG9TgzKj0hu5St-r1-174:-L4J532-Jfgvogi5MGXO:bfb4jfu) ([raw file](https://github.com/pair-code/deeplearnjs/blob/42d15710993c81e9af9bca5e81011203d4b245a9/src/math/backends/backend_engine.ts#L174)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

we should figure out terminology here.

either 1) tidy is the only terminology in the code base or 2) scope is the only terminology in the core codebase, and tidy is simply the API-surface alias.

Speaking offline, lets go with 2)
</blockquote></details>

Went with 2)

---

*[src/math/backends/gradients.ts, line 37 at r1](https://beta.reviewable.io/reviews/pair-code/deeplearnjs/625#-L4D5Feq_E0VzsdVn7UA-r1-37:-L4J573GSJL_CAGnNRo8:b11476b) ([raw file](https://github.com/pair-code/deeplearnjs/blob/42d15710993c81e9af9bca5e81011203d4b245a9/src/math/backends/gradients.ts#L37)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

here we have "Scope" - see previous comment about "tidy" vs "scope" terminology
</blockquote></details>

Went with scope internally. Also I didn't export this method to public API (not yet) since we'll do the auto-grad style `grad(f)(x)`, at which point there might not be any need for this method to be user-facing.

---

*[src/math/backends/gradients.ts, line 58 at r1](https://beta.reviewable.io/reviews/pair-code/deeplearnjs/625#-L4D5Feq_E0VzsdVn7UA-r1-58:-L4NlwDPEkyuTkA4mZtq:b-896fix) ([raw file](https://github.com/pair-code/deeplearnjs/blob/42d15710993c81e9af9bca5e81011203d4b245a9/src/math/backends/gradients.ts#L58)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

let's push here and below into engine
</blockquote></details>

Done.

---

*[src/math/backends/gradients.ts, line 125 at r1](https://beta.reviewable.io/reviews/pair-code/deeplearnjs/625#-L4D5Feq_E0VzsdVn7UA-r1-125:-L4Nlwgt9esLn9mVupxv:b-896fix) ([raw file](https://github.com/pair-code/deeplearnjs/blob/42d15710993c81e9af9bca5e81011203d4b245a9/src/math/backends/gradients.ts#L125)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

let's push here and below into the engine. this in-function input computation should be hidden in the sugar floating function
</blockquote></details>

Done.

---

*[src/math/backends/gradients.ts, line 199 at r1](https://beta.reviewable.io/reviews/pair-code/deeplearnjs/625#-L4D5Feq_E0VzsdVn7UA-r1-199:-L4JBaKklLZFP_7R3RkN:b-896fix) ([raw file](https://github.com/pair-code/deeplearnjs/blob/42d15710993c81e9af9bca5e81011203d4b245a9/src/math/backends/gradients.ts#L199)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

let's push all of this into the engine
</blockquote></details>

Done.

---

*[src/math/backends/tape_util.ts, line 244 at r1](https://beta.reviewable.io/reviews/pair-code/deeplearnjs/625#-L4E65uHbAzAncZe7wMi-r1-244:-L4J7LKEOzlq1tmaoI_V:b8k3ipi) ([raw file](https://github.com/pair-code/deeplearnjs/blob/42d15710993c81e9af9bca5e81011203d4b245a9/src/math/backends/tape_util.ts#L244)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

extractNDArraysFromTidyResult
</blockquote></details>

went with scope internally, so this stays the same

---

*[models/knn_image_classifier/knn_image_classifier.ts, line 88 at r1](https://beta.reviewable.io/reviews/pair-code/deeplearnjs/625#-L4DyQmN-cpAKxEFZK53-r1-88:-L4J2b1nQSb9K6FcSyt7:bj9l1ny) ([raw file](https://github.com/pair-code/deeplearnjs/blob/42d15710993c81e9af9bca5e81011203d4b245a9/models/knn_image_classifier/knn_image_classifier.ts#L88)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

dl.tidy

BTW - can we just not export tidy on math?
</blockquote></details>

Removed math.tidy. Reverting changes in models/* since they depend on older version of dljs and should be updated in a separate PR.

---

*[models/knn_image_classifier/knn_image_classifier.ts, line 131 at r1](https://beta.reviewable.io/reviews/pair-code/deeplearnjs/625#-L4DyQmN-cpAKxEFZK53-r1-131:-L4J3EAuyQTYo7Br60Y2:b-tue9kl) ([raw file](https://github.com/pair-code/deeplearnjs/blob/42d15710993c81e9af9bca5e81011203d4b245a9/models/knn_image_classifier/knn_image_classifier.ts#L131)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

dl.tidy
</blockquote></details>

Reverted changes in models/*

---

*[models/mobilenet/mobilenet.ts, line 53 at r1](https://beta.reviewable.io/reviews/pair-code/deeplearnjs/625#-L4DyQmN-cpAKxEFZK54-r1-53:-L4J6b7zfv5vhpOcKAmh:b-tue9kl) ([raw file](https://github.com/pair-code/deeplearnjs/blob/42d15710993c81e9af9bca5e81011203d4b245a9/models/mobilenet/mobilenet.ts#L53)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

dl.tidy
</blockquote></details>

Reverted changes in models/*

---

*[models/squeezenet/squeezenet.ts, line 66 at r1](https://beta.reviewable.io/reviews/pair-code/deeplearnjs/625#-L4DyQmN-cpAKxEFZK55-r1-66:-L4J2sSbab0e3YwKJUYs:b-vi17hg) ([raw file](https://github.com/pair-code/deeplearnjs/blob/42d15710993c81e9af9bca5e81011203d4b245a9/models/squeezenet/squeezenet.ts#L66)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

dl.tidy
</blockquote></details>

Reverting changes in models/*

---

*[models/squeezenet/squeezenet.ts, line 175 at r1](https://beta.reviewable.io/reviews/pair-code/deeplearnjs/625#-L4DyQmN-cpAKxEFZK55-r1-175:-L4J2ueHCaJfah3D3Tid:b-vi17hg) ([raw file](https://github.com/pair-code/deeplearnjs/blob/42d15710993c81e9af9bca5e81011203d4b245a9/models/squeezenet/squeezenet.ts#L175)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

dl.tidy
</blockquote></details>

Reverting changes in models/*

---

*[models/yolo_mobilenet/yolo_mobilenet.ts, line 58 at r1](https://beta.reviewable.io/reviews/pair-code/deeplearnjs/625#-L4DyQmN-cpAKxEFZK56-r1-58:-L4J36dPjnzG5WHHraia:b-tue9kl) ([raw file](https://github.com/pair-code/deeplearnjs/blob/42d15710993c81e9af9bca5e81011203d4b245a9/models/yolo_mobilenet/yolo_mobilenet.ts#L58)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

dl.tidy
</blockquote></details>

Reverted changes in models/*

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/pair-code/deeplearnjs/625)*
<!-- Sent from Reviewable.io -->

nsthorat commented:
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/>

---

Reviewed 29 of 74 files at r1, 50 of 50 files at r2.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[demos/mnist/mnist.ts, line 29 at r2](https://reviewable.io:443/reviews/pair-code/deeplearnjs/625#-L4R92uMqKVXoKpNhtjH:-L4R92uMqKVXoKpNhtjI:baho6z1) ([raw file](https://github.com/pair-code/deeplearnjs/blob/f5e1d0c16debc32657dd2d84621a8d8a96cfe7cb/demos/mnist/mnist.ts#L29)):*
> ```XML
>     const data = JSON.parse(xhr.responseText) as SampleData;
> 
>     // Wrap everything in a math.scope so we clean up intermediate NDArrays.
> ```

update comment

---

*[docs/roadmap.md, line 20 at r2](https://reviewable.io:443/reviews/pair-code/deeplearnjs/625#-L4R9DdMmkKFXQpXXCsV:-L4R9DdMmkKFXQpXXCsW:b-jvbi79) ([raw file](https://github.com/pair-code/deeplearnjs/blob/73ef966b7ba048ed9e7172da105c1fb2f2f6cf6b/docs/roadmap.md#L20)):*
> ```Markdown
> **deeplearn.js** automatically from a `GraphDef`.
> 
> ## Decoupling NDArray from storage mechanism
> ```

you can remove this whole section actually

---

*[src/math/backends/gradients.ts, line 67 at r2](https://reviewable.io:443/reviews/pair-code/deeplearnjs/625#-L4RDLy8R-NUSzi4FK0L:-L4RDLy8R-NUSzi4FK0M:bg4wtt9) ([raw file](https://github.com/pair-code/deeplearnjs/blob/73ef966b7ba048ed9e7172da105c1fb2f2f6cf6b/src/math/backends/gradients.ts#L67)):*
> ```XML
> export function gradients<R extends Rank, T extends NDArray|NamedArrayMap>(
>     f: () => NDArray<R>, x: T): T {
>   return valueAndGradients(f, x).gradients;
> ```

Hold onto this result and dispose the value

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pair-code/deeplearnjs/625#-:-L4REYX1ymG_bmU7W3vq:b-jbsf8z)*
<!-- Sent from Reviewable.io -->

dsmilkov commented:
Review status: 72 of 75 files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[demos/mnist/mnist.ts, line 29 at r2](https://reviewable.io:443/reviews/pair-code/deeplearnjs/625#-L4R92uMqKVXoKpNhtjH:-L4RGWyJ9Wj5lhE92c6X:b-896fix) ([raw file](https://github.com/pair-code/deeplearnjs/blob/f5e1d0c16debc32657dd2d84621a8d8a96cfe7cb/demos/mnist/mnist.ts#L29)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

update comment
</blockquote></details>

Done.

---

*[docs/roadmap.md, line 20 at r2](https://reviewable.io:443/reviews/pair-code/deeplearnjs/625#-L4R9DdMmkKFXQpXXCsV:-L4RGXpmFGtNMA0fu8Zb:b-896fix) ([raw file](https://github.com/pair-code/deeplearnjs/blob/73ef966b7ba048ed9e7172da105c1fb2f2f6cf6b/docs/roadmap.md#L20)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

you can remove this whole section actually
</blockquote></details>

Done.

---

*[src/math/backends/gradients.ts, line 67 at r2](https://reviewable.io:443/reviews/pair-code/deeplearnjs/625#-L4RDLy8R-NUSzi4FK0L:-L4RGY_ZNedjXxEuJmsD:b-896fix) ([raw file](https://github.com/pair-code/deeplearnjs/blob/73ef966b7ba048ed9e7172da105c1fb2f2f6cf6b/src/math/backends/gradients.ts#L67)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

Hold onto this result and dispose the value
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pair-code/deeplearnjs/625)*
<!-- Sent from Reviewable.io -->
